src/workerd/api/messagechannel.c++

```diff
@@ -0,0 +1,317 @@
+}
+
+MessagePort::MessagePort(IoContext& ioContext): ioContext(ioContext), state(Pending()) {
+  // We set a callback on the underlying EventTarget to be notified when
+  // a listener for the message event is added or removed. When there
+  // are no listeners, we move back to the Pending state, otherwise we
+  // will switch to the Started state if necessary.
+  setEventListenerCallback([&](jsg::Lock& js, kj::StringPtr name, size_t count) {
+    if (name == "message"_kj) {
+      KJ_SWITCH_ONEOF(state) {
+        KJ_CASE_ONEOF(pending, Pending) {
+          // If we are in the pending state, start the port if we have listeners.
+          // This is technically not spec compliant, but it is what Node.js
+          // supports. Specifically, adding a new message listener using the
+          // addEventListener method is *technically* not supposed to start
+          // the port but we're going to do what Node.js does.
+          if (count > 0 || onmessageValue != kj::none) {
+            start(js);
+          }
+        }
+        KJ_CASE_ONEOF(started, Started) {
+          // If we are in the started state, stop the port if there are no listeners.
+          if (count == 0 && onmessageValue == kj::none) {
+            state = Pending();
+          }
+        }
+        KJ_CASE_ONEOF(_, Closed) {
+          // Nothing to do. We're already closed so we don't care.
+        }
+      }
+    }
+  });
+}
+
+// Deliver the message to the "message" or "messageerror" event on this port.
+void MessagePort::deliver(
+    jsg::Lock& js, kj::Own<rpc::JsValue::Reader> message, jsg::Ref<MessagePort> port) {
+  auto event = js.tryCatch([&] {
+    jsg::Deserializer deserializer(js, message->getV8Serialized());
+    return MessageEvent::New(
+        js, jsg::JsRef(js, deserializer.readValue(js)), kj::str(), port->addRef(), {});
+  }, [&](jsg::Value exception) {
+    return MessageEvent::NewError(
+        js, jsg::JsRef(js, jsg::JsValue(exception.getHandle(js))), kj::str(), port->addRef(), {});
+  });
+
+  // If any of the message/messageerror event handlers throw,
+  // capture the error and pass it to reportError instead of
+  // propagating up.
+  js.tryCatch([&] { port->dispatchEventImpl(js, kj::mv(event)); }, [&](jsg::Value exception) {
+    auto context = js.v8Context();
+    auto& global =
+        jsg::extractInternalPointer<ServiceWorkerGlobalScope, true>(context, context->Global());
+    global.reportError(js, jsg::JsValue(exception.getHandle(js)));
+  });
+}
+
+// Deliver the message to all the jsrpc remotes we have
+kj::Promise<void> MessagePort::sendToRpc(kj::Own<rpc::JsValue::Reader> message) {
+  KJ_IF_SOME(outputLocks, ioContext.waitForOutputLocksIfNecessary()) {
+    co_await outputLocks;
+  }
+  kj::Vector<kj::Promise<void>> promises;
+  for (rpc::JsMessagePort::Client cap: rpcClients) {
+    auto req = cap.callRequest();
+    req.setData(*message);
+    // TODO(message-port): Removing the port when dropped?
```

Comment:
You're referencing to the case where MessagePort is removed and the existing promises needs to be cancelled right?

Reply:
no, if the capability is dropped on the remote side then we should drop the local one too. Right now I think this leaks. Will need to resolve this before this is ready to land.